### DISCUSSION TITLE
Generation of variables from running a script

### DISCUSSION THREAD

**Discussion by darewreck54** [Huachao/vscode-restclient#869] [General]
In my use case, I would like to generate an auth token and inject it into a variable so that variable can be used in the request.

````
@token = $(bash getToken.sh)

GET https://example.com
HTTP/1.1
Token: {{token}}
````

Is there currently anyway to support this?  I imagine when it look for the variable it would evaluate it during runtime.

**Comment by khera** [comment]
I have a similar need where I have to compute a token with multiple comma separate values and also include an SHA1 hash of those variables. Something like this:

    Token: a="1", b="2", c="3", hash=Base64(SHA1("1" + "2" + "3"))

I really would love a way to eval some function to produce the token on every call. The values are not static: one is a timestamp and one is a nonce that changes per request. I was thinking something like a `pre_request_hook` I can define, or some syntax like `{{generate_token()}}` to call my JS function defined in the settings file.

**Comment by MominRaza** [comment]
Related update: https://github.com/Huachao/vscode-restclient/pull/1422

This discussion is addressed by the new `@set` directive, which lets you capture response values (`response.headers.*` / `response.body.*`) into shared runtime variables and reuse them across `.http` files.

**Reply by khera** [reply]
I don't follow how these are related.

We need a way to call a local function or shell command and save that value for use later, similarly to how you would capture the output of a REST call.

**Reply by gorillamoe** [reply]
Kulala for (Visual Studio) Code finally landed about a week ago.

We have a toggle to enable basic compatibility mode for this extension and also support all the Jetbrains http client features (plus additional features).

It also comes with full tree-sitter syntax highlighting with injections and its own language server.

We support all `curl` flags (short and long name) and have full` jq`support for filtering if response bodies.

Since we have extended scripting capabilities, there are literally no limit on what you can do.

We have examples for shelling out to keepassxc and 1password to fetch passwords, but that's just to give you a glimpse of what is possible.

You may want to give it a whirl: 

- https://marketplace.visualstudio.com/items?itemName=mistweaverco.kulala
- https://open-vsx.org/extension/mistweaverco/kulala

Feedback would be highly appreciated 👍🏾
